pkg/k8sutil/sort.go

```diff
@@ -0,0 +1,177 @@
+			return false
+		}
+		if !versionRegexp.MatchString(verJ) {
+			return true
+		}
+		subsI := versionRegexp.FindStringSubmatch(verI)
+		subsJ := versionRegexp.FindStringSubmatch(verJ)
+		subsI, subsJ = removeEmpties(subsI), removeEmpties(subsJ)
+		// Two single grouping, ex. "(v1)" vs "(v2)"
+		if len(subsI) == 1 && len(subsI) == len(subsJ) {
+			return parseInt(subsI[0]) > parseInt(subsJ[0])
+		}
+		// Two quad groupings, ex. "(v1)((alpha)(1))" vs "(v2)((alpha)(1))"
+		if len(subsI) == 4 && len(subsI) == len(subsJ) {
+			// If both alpha or beta, sort by version numbers.
+			if subsI[2] == subsJ[2] {
+				if subsI[0] == subsJ[0] {
+					return parseInt(subsI[3]) > parseInt(subsJ[3])
+				}
+				return parseInt(subsI[0]) > parseInt(subsJ[0])
+			}
+			// If I isn't beta then J must be.
+			return subsI[2] == "beta"
+		}
+		// A single and a quad grouping. Always sort single before quad.
+		return len(subsI) == 1
+	}
+}
+
+// SortVersionsMajor sorts CRD versions by their leading major version.
+// Example:
+//	input:  ["v1alpha1", "v1", "v2beta4", "v2alpha4", "v3alpha1", "v2"]
+//	output: ["v3alpha1", "v2", "v2beta4", "v2alpha4", "v1", "v1alpha1"]
+func SortVersionsMajor(versions interface{}, getVer GetVerFunc) {
+	less := sortVersionsMajorFunc(versions, getVer)
+	runSort(versions, less)
+}
+
+func sortVersionsMajorFunc(versions interface{}, getVer GetVerFunc) func(int, int) bool {
+	return func(i int, j int) bool {
+		verI, verJ := getVer(versions, i), getVer(versions, j)
+		if !versionRegexp.MatchString(verI) && !versionRegexp.MatchString(verJ) {
+			return verI < verJ
+		}
+		if !versionRegexp.MatchString(verI) {
+			return false
+		}
+		if !versionRegexp.MatchString(verJ) {
+			return true
+		}
+		subsI := versionRegexp.FindStringSubmatch(verI)
+		subsJ := versionRegexp.FindStringSubmatch(verJ)
+		subsI, subsJ = removeEmpties(subsI), removeEmpties(subsJ)
+		// Two single grouping, ex. "(v1)" vs "(v2)"
+		if len(subsI) == 1 && len(subsI) == len(subsJ) {
+			return subsI[0] > subsJ[0]
+		}
+		// Two quad groupings, ex. "(v1)((alpha)(1))" vs "(v2)((alpha)(1))"
+		if len(subsI) == 4 && len(subsI) == len(subsJ) {
+			if subsI[0] == subsJ[0] {
+				if subsI[2] == subsJ[2] {
+					return parseInt(subsI[3]) > parseInt(subsJ[3])
+				}
+				// If I isn't beta then J must be.
+				return subsI[2] == "beta"
+			}
+			return parseInt(subsI[0]) > parseInt(subsJ[0])
+		}
+		// A single and a quad grouping. Either I and J have equal major groups
+		// and I is stable, visa versa, or one has a greater major group.
+		firstGroupEqual := subsI[0] == subsJ[0]
+		if len(subsI) == 1 {
+			return firstGroupEqual || parseInt(subsI[0]) > parseInt(subsJ[0])
+		}
+		return !firstGroupEqual && parseInt(subsI[0]) >= parseInt(subsJ[0])
+	}
+}
+
+// GetCRDVersionsName gets the string value from v[i].Name. GetCRDVersionsName
+// panics if v is not a []v1beta1.CustomResourceDefinitionVersion.
+func GetCRDVersionsName(v interface{}, i int) string {
```

Comment:
To avoid runtime panics, I think I'd prefer functions that convert concrete types into `[]string`, which can be passed to `SortVersions` as `[]string` rather than interface, which also removes the need for passing an additional getter function. WDYT?

Reply:
I use `interface{}` so `SortVersions` can sort a slice of arbitrary data structures by a version key, ex. `v1beta1.CustomResourceDefinitionVersion`. I am not a fan of runtime panics either; let me think about a better way to implement version-getting.

Reply:
@joelanford there are two solutions I can think of:

1. Use a `chan` and `sync.WaitGroup` to receive an error from a `GetVerFunc`.
1. Add a simple check to `SortVersions*`:
```
func SortVersions(versions interface{}, getVer GetVerFunc) error {
	if reflect.ValueOf(versions).Len() == 0 {
		return nil
	}
	if _, err := getVer(versions, 0); err != nil {
		return err
	}
	...
}
```

1 is more complex and messy, and while more robust than 2 I like 2's simplicity. WDYT?

Reply:
You definitely do not need/want a channel or synchronization here.

Reply:
@chancez correct there's no need for synchronization here. The `chan` would be to get the error out of the `Less()` function, and the `sync.WaitGroup` is necessary to avoid `chan` blocking (`runSort` would be in its own goroutine).

As for 2, yes you're right however `SortVersions` can assume `versions` is a slice type if we go with 2. This isn't explicit in the function comment though. I'll update it if we go with 2.

Reply:
@estroz Oh gotcha. I missed the sorting of arbitrary types by key the first time around. Don't know if this would work, but what about something like:

```go
type StringVersions struct {
    versions []string
}

func (vs StringVersions) Len() int { ... }
func (vs StringVersions) Less(i, j int) bool { return lessK8sVersion(vs[i], vs[j]) }
func (vs StringVersions) Swap(i, j int) bool { ... }

type CRDVersions struct {
    versions []apiextv1beta1.CustomResourceDefinitionVersion
}

func (vs StringVersions) Len() int { ... }
func (vs StringVersions) Less(i, j int) bool { return lessK8sVersion(vs[i].Name, vs[j].Name) }
func (vs StringVersions) Swap(i, j int) bool { ... }

func main() {
    sliceA := []string{"v1", "v2beta2", "v1alpha1"}
    sliceB := []apiextv1beta1.CustomResourceDefinitionVersion{b,a,d,c}
    strVersions := StringVersions{Versions: sliceA}
    crdVersions := CRDVersions{Versions: sliceB}

    sort.Sort(strVersions)
    sort.Sort(crdVersions)
}
```

Would that modify the underlying slice data?

Reply:
> @chancez correct there's no need for synchronization here. The `chan` would be to get the error out of the `Less()` function, and the `sync.WaitGroup` is necessary to avoid `chan` blocking (`runSort` would be in its own goroutine).
You can use a callback for that.

Reply:
@chancez can you give an example of a callback in this context? It doesn't necessarily matter anymore (@joelanford's suggestion is a way better interface so I'm going with that), but I'd like to know for my own knowledge.

Reply:
```

Reply:
Your 2nd example seems like it was basically what I meant.